src/puppetlabs/trapperkeeper/core.clj

```diff
@@ -139,29 +145,31 @@
   Blocks the calling thread until trapperkeeper is shut down.
   `cli-data` is expected to be a map constructed by parsing the CLI args.
   (see `parse-cli-args`)"
-  [cli-data]
-  {:pre [(map? cli-data)]}
-  (let [app (boot-with-cli-data cli-data)]
-    ;; This line populates the `main-app` variable with the TrapperkeeperApp
-    ;; instance.  This allows it to be referenced in a remote nREPL session.
-    (alter-var-root #'main-app (fn [_] app))
-    (run-app app)))
-
-(defn main
-  "Launches the trapperkeeper framework. This function blocks until
-  trapperkeeper is shut down. This may be called directly, but is also called by
-  `puppetlabs.trapperkeeper.core/-main` if you use `puppetlabs.trapperkeeper.core`
-  as the `:main` namespace in your leinengen project."
-  [& args]
+  ([cli-data preprocess-config-fn]
+   {:pre [(map? cli-data)
+          (fn? preprocess-config-fn)]}
+   (let [app (boot-with-cli-data cli-data preprocess-config-fn)]
+     ;; This line populates the `main-app` variable with the TrapperkeeperApp
+     ;; instance.  This allows it to be referenced in a remote nREPL session.
+     (alter-var-root #'main-app (fn [_] app))
+     (run-app app)))
+  ([cli-data] (run cli-data identity)))
+
+(defn main-with-config-hook
```

Comment:
Just to make sure I understand this, an example of it's usage would be PuppetDB having it's own main function:

``` clojure
(defn our-real-main [& args]
  (main-with-config-hook args pdb-specific-config-fn))
```
Then our tests would switch to the new main function and we would update our EZBake config to include this new main-namespace by setting a [config propery ](https://github.com/puppetlabs/ezbake/blob/master/src/puppetlabs/ezbake/core.clj#L405).

When I created the defaulted config service in PuppetDB I had a much uglier hack that I [pushed up here](https://github.com/senior/trapperkeeper/commit/c09bef628817257a9ae847e5d8dfdd3d058c59cb). I think this approach (if I'm understanding it correctly) is more work, but is better and it's much more obvious what's going on.